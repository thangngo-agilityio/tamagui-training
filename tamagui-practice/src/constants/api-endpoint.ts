export const BASE_URL = process.env.NEXT_PUBLIC_BASE_URL || '';
export const SECOND_URL = process.env.NEXT_PUBLIC_SECOND_URL || '';

export const API_PATH = {
  USERS: 'users',
  PRODUCTS: 'products',
  CARTS: 'cart',
};
