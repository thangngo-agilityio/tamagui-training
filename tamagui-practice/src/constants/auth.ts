export const AUTH_METHODS = {
  CREDENTIALS: 'credentials',
};

export const ERROR_TYPES = {
  CREDENTIALS_SIGN_IN: 'CredentialsSignin',
};
