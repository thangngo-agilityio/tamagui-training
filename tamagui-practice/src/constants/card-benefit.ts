import { CheckIcon, CupIcon, ShippingIcon, SupportIcon } from '@/icons';

export const BENEFIT_LIST = [
  {
    id: '1',
    icon: CupIcon,
    title: 'High Quality',
    text: 'crafted from top materials',
  },
  {
    id: '2',
    icon: CheckIcon,
    title: 'Warrany Protection',
    text: 'Over 8 Months',
  },
  {
    id: '3',
    icon: ShippingIcon,
    title: 'Free Shipping',
    text: 'Order over N300,000',
  },
  {
    id: '4',
    icon: SupportIcon,
    title: '24 / 7 Support',
    text: 'Dedicated support',
  },
];
