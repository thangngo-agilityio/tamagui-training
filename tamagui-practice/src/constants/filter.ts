import {
  BedActiveIcon,
  BedIcon,
  ChairActiveIcon,
  ChairIcon,
  SofaActiveIcon,
  SofaIcon,
  StarActiveIcon,
  StarIcon,
  TableActiveIcon,
  TableIcon,
} from '@/icons';

export const MENU_ITEM_FILTER = [
  {
    id: '1',
    icon: StarIcon,
    iconActive: StarActiveIcon,
    itemContent: 'Popular',
    value: '',
  },
  {
    id: '2',
    icon: ChairIcon,
    iconActive: ChairActiveIcon,
    itemContent: 'Chair',
    value: 'chair',
  },
  {
    id: '3',
    icon: TableIcon,
    iconActive: TableActiveIcon,
    itemContent: 'Workstation',
    value: 'workstation',
  },
  {
    id: '4',
    icon: SofaIcon,
    iconActive: SofaActiveIcon,
    itemContent: 'Living room',
    value: 'living room',
  },
  {
    id: '5',
    icon: BedIcon,
    iconActive: BedActiveIcon,
    itemContent: 'Bedroom',
    value: 'bedroom',
  },
];
