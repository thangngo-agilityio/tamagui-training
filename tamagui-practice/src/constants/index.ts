export * from './times';
export * from './router';
export * from './api-endpoint';
export * from './form';
export * from './filter';
export * from './card-benefit';
export * from './footer';
export * from './auth';
export * from './message';
export * from './pagination';
export * from './navigation';
export * from './query-tags';
export * from './regex';
