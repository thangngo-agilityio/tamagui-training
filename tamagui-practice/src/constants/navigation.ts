export const SEARCH_QUERIES = {
  NAME: 'name',
  CATEGORIES: 'category',
};
