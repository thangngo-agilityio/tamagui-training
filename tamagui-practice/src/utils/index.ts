export * from './images';
export * from './pagination';
export * from './paramUrl';
export * from './helper';
export * from './cart';
export * from './format';
export * from './validators';
