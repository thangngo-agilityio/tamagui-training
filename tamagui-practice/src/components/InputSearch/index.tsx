'use client';

import { Input, InputGroup, InputLeftElement } from '@chakra-ui/react';
import { usePathname, useRouter, useSearchParams } from 'next/navigation';
import { useDebounceCallback } from 'usehooks-ts';

// Constants
import { ROUTER, SEARCH_QUERIES } from '@/constants';

// Icons
import { SearchIcon } from '@/icons';

// Utils
import { getSearchParams, updateSearchParams } from '@/utils';

const InputSearch = () => {
  const searchParams = useSearchParams();
  const pathname = usePathname();
  const { replace } = useRouter();
  const { name } = getSearchParams(searchParams);

  const isShowSearch = pathname === ROUTER.PRODUCT;

  const handleOnChange = useDebounceCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      const value = e.target.value;
      const updatedParams = updateSearchParams(
        searchParams,
        SEARCH_QUERIES.NAME,
        value,
      );

      replace(`${pathname}?${updatedParams.toString()}`, { scroll: false });
    },
    500,
  );
  return (
    <>
      {isShowSearch && (
        <InputGroup>
          <InputLeftElement>
            <SearchIcon />
          </InputLeftElement>
          <Input
            placeholder="Search for minimalist chair"
            defaultValue={name}
            onChange={handleOnChange}
            variant="search"
            background="background.100"
            data-testid="search"
          />
        </InputGroup>
      )}
    </>
  );
};

export default InputSearch;
