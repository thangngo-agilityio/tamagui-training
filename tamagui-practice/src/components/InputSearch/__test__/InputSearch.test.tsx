import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { usePathname, useRouter, useSearchParams } from 'next/navigation';
import '@testing-library/jest-dom';

// Components
import InputSearch from '../index';

// Constants
import { SEARCH_QUERIES } from '@/constants';

// Utils
import { getSearchParams, updateSearchParams } from '@/utils';

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
  usePathname: jest.fn(),
  useSearchParams: jest.fn(),
}));

jest.mock('@/utils', () => ({
  getSearchParams: jest.fn(),
  updateSearchParams: jest.fn(),
}));

jest.mock('@/icons', () => ({
  SearchIcon: () => <svg data-testid="search-icon" />,
}));

describe('InputSearch component', () => {
  let replaceMock: jest.Mock;
  let searchParamsMock: URLSearchParams;
  let pathnameMock: string;

  beforeEach(() => {
    replaceMock = jest.fn();
    searchParamsMock = new URLSearchParams();
    pathnameMock = '/test-path';

    (useRouter as jest.Mock).mockReturnValue({ replace: replaceMock });
    (usePathname as jest.Mock).mockReturnValue(pathnameMock);
    (useSearchParams as jest.Mock).mockReturnValue(searchParamsMock);
    (getSearchParams as jest.Mock).mockReturnValue({ name: '' });
    (updateSearchParams as jest.Mock).mockReturnValue(new URLSearchParams());
  });

  it('should match snapshot for InputSearch ', () => {
    const { container } = render(<InputSearch />);

    expect(container).toMatchSnapshot();
  });

  describe('Happy Path', () => {
    it('should render the input with default placeholder and no initial value', () => {
      // Test to ensure the input renders correctly with default settings
      render(<InputSearch />);
      const inputElement = screen.getByPlaceholderText(
        'Search for minimalist chair',
      );
      expect(inputElement).toBeInTheDocument();
      expect(inputElement).toHaveValue('');
    });

    it('should update the search params and call replace on input change', () => {
      // Test to ensure input change updates search params and calls replace
      const updatedParams = new URLSearchParams('name=test');
      (updateSearchParams as jest.Mock).mockReturnValue(updatedParams);

      render(<InputSearch />);
      const inputElement = screen.getByPlaceholderText(
        'Search for minimalist chair',
      );
      fireEvent.change(inputElement, { target: { value: 'test' } });

      setTimeout(() => {
        expect(updateSearchParams).toHaveBeenCalledWith(
          searchParamsMock,
          SEARCH_QUERIES.NAME,
          'test',
        );
        expect(replaceMock).toHaveBeenCalledWith(
          `${pathnameMock}?${updatedParams.toString()}`,
          { scroll: false },
        );
      }, 500);
    });
  });

  describe('Edge Cases', () => {
    it('should handle empty input gracefully', () => {
      // Test to ensure empty input does not cause errors
      render(<InputSearch />);
      const inputElement = screen.getByPlaceholderText(
        'Search for minimalist chair',
      );
      fireEvent.change(inputElement, { target: { value: '' } });

      setTimeout(() => {
        expect(updateSearchParams).toHaveBeenCalledWith(
          searchParamsMock,
          SEARCH_QUERIES.NAME,
          '',
        );
        expect(replaceMock).toHaveBeenCalledWith(`${pathnameMock}?`, {
          scroll: false,
        });
      }, 500);
    });

    it('should debounce input changes', () => {
      // Test to ensure input changes are debounced
      jest.useFakeTimers();
      render(<InputSearch />);
      const inputElement = screen.getByPlaceholderText(
        'Search for minimalist chair',
      );
      fireEvent.change(inputElement, { target: { value: 'a' } });
      fireEvent.change(inputElement, { target: { value: 'ab' } });
      fireEvent.change(inputElement, { target: { value: 'abc' } });

      jest.advanceTimersByTime(500);

      expect(updateSearchParams).toHaveBeenCalledTimes(1);
      expect(replaceMock).toHaveBeenCalledTimes(1);

      jest.useRealTimers();
    });
  });
});
