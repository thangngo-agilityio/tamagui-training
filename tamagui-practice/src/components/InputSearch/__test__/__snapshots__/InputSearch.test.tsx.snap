// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputSearch component should match snapshot for InputSearch  1`] = `
<div>
  <div
    class="chakra-input__group css-bx0blc"
    data-group="true"
  >
    <div
      class="chakra-input__left-element css-o84xzr"
    >
      <svg
        data-testid="search-icon"
      />
    </div>
    <input
      class="chakra-input css-q179pw"
      data-testid="search"
      placeholder="Search for minimalist chair"
      value=""
    />
  </div>
</div>
`;
